File: chaeyun/1_permute-combin/subset-dup-input/subset-dup.cpp
class Solution {
public:
    void fun(vector<int>& nums, int index, vector<int>ds, set<vector<int>>& res) {
        if(index == nums.size()) {
            res.insert(ds);
            return;
        }
        ds.push_back(nums[index]);
        fun(nums, index+1, ds, res);
        ds.pop_back();
        fun(nums, index+1, ds, res);
    }
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        set<vector<int>> res;
        vector<int> ds;
        fun(nums, 0, ds, res);
        for(auto it = res.begin(); it!= res.end(); it++) {
            ans.push_back(*it);
        }  
        return ans;
    }
};

File: chaeyun/1_permute-combin/subset-dup-input/subset-dup.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        ans = []
        res = set()
        ds = []
        nums.sort() 
        self.fun(nums, 0, ds, res)

        for subset in res:
            ans.append(list(subset))

        return ans

    def fun(self, nums, index, ds, res):
        if index == len(nums):
            sorted_ds = tuple(sorted(ds))
            res.add(sorted_ds)
            return

        ds.append(nums[index])
        self.fun(nums, index + 1, ds, res)
        ds.pop()
        self.fun(nums, index + 1, ds, res)
